[Kotlin] #1716. Add implementation for concept exercise: `basic` (#1831)
exercism/v3: languages/kotlin/exercises/concept/basics/.docs/introduction.md

Kotlin is **statically typed** programming language developed by JetBrains. This means that type of variables is defined on compile-time.

## Entry point

To run your Kotlin program you need to define so-called entry point: function with name `main` with or without arguments. However, in Exercism course we are using automatic tests that are using other functions defined in exercises.

Entry point with arguments (you can use them for building CLI applications):

```kotlin
fun main(args: Array<String>) {
    println("Hello, Exercism!")
    println("Program args are: $args")
}
```

Or ignore arguments completely:

```kotlin
fun main() {
    println("Hello, Exercism!")
}

> ```

Reviewer: Is this information needed for the student to solve the exercise? I suspect it isn't, in which case this information should be removed. In general, the `introduction.md` document should only contain information that is needed to understand the fundamentals of the concept and solve the exercise should be provided (see [this description](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd)). You could consider having this information in the `after.md` document though.

Author: You are right. It's not for exercise but it's kinda important thing to know to be able to say that language is "mastered". Moving it to `after` is a great idea. Thanks.

print(index)  // 12

index = 100
print(index)  // 100
```

BTW probably you have already noticed that semicolons in Kotlin are not used (they are optional but not needed except few special cases we will explore later in other exercises).

## Functions

Functions in Kotlin are defined with keyword **`fun`** and are [first-class citizens][wiki-fcc] (not related to OOP). It means that you can declare (so-called `top-level functions`) them right in files (e.g. in Java you can define methods only in classes, not in files):

```kotlin
// This is content of the Hello.kt file

fun hello() {}
```

Functions can receive arguments. Each argument have a name and a type. Probably you remember that compiler is smart enough to understand type of variable In many cases, but you always need to write types for function arguments by yourself as you are declaring certain contract. Functions can have zero or more arguments:

```kotlin
fun hello() {}

fun hello(name: String) {}

fun hello(name: String, age: Int) {}
```

Kotlin functions always return value. Either explicitly (with specified type):

```kotlin
fun min(a: Int, b: Int): Int

fun countBonuses(user: User): Bonuses

fun run(): Unit {}
```

> Or implicitly. In this case return type will be `Unit` (similar to `void` from C/Java) and can be omitted:

Reviewer: Is the `Unit` concept needed to solve the exercise? If not, you could consider postponing introducing this Concept to a separate exercise (in C#, we introduce it in the `classes` exercise).

Author: Should I move this to `after` too? `Unit` is quite basic stuff and not really connected to classes.

Reviewer: Yeah sure, you can move it to after.

```kotlin
fun run() {}

// is the same as
fun run(): Unit { return Unit }
```

As you can see, to return value from function you need to use keyword **`return`**:

```kotlin
fun getName(): String {
    return "Alice"
}
```

Functions can have parameters with default values. These values will be used if they are omitted where function is invoked:

```kotlin
fun ping(host: String = "localhost") {
    println("PING --> $host")
}

ping("exercism.io")  // PING --> exercism.io

> ping()               // PING --> localhost

Reviewer: Are default values needed to solve this exercise?

Author: Yeah, I will add reference implementation.

```

## Comments

Use `//` to define single-line comment:

```kotlin
foo() // Everything after `//` will be ignored by compiler

// I will be ignored too
```

or `/*` and `*/` to define multi-line comments:

```kotlin
/*
   This this an example
   for a multiline comment
*/
```

> ## Running tests

Author: Ah, I see. Will remove this from introduction.

To check your solution - use `./gradlew test` in console (in GNU/Linux. MacOS) or `gradlew.bat test` (in Windows).

> If you are using [IntelliJ IDEA (Community Edition)][intellij-idea-ic] - use `Run Tests` from your toolbar.

Author: Will be prepared and added to git

> Kotlin is **statically typed** programming language developed by JetBrains. This means that type of variables is defined on compile-time.

Reviewer: ```suggestion
Kotlin is a **statically typed** programming language developed by JetBrains. This means that the type of variables is defined at compile-time.
```

## Variables

> Similarly to other statically-typed programming languages type of each variable should be defined in compile time. Similarly to other **modern** programming languages you can avoid explicit type declarations where they can be interfered by compiler from context. Use this ability to make your code more readable. But remember that there are cases where it's better to specify type explicitly - e.g. in public APIs.

Reviewer: ```suggestion
Similarly to other statically-typed programming languages, the type of each variable should be defined at compile time. You can avoid explicit type declarations where they can be inferred by the compiler from their context.
```

I've suggested removing the following:

> Use this ability to make your code more readable. But remember that there are cases where it's better to specify type explicitly - e.g. in public APIs.

The reason for this is to keep the introduction as concise as possible. The introduction should only contain information that is needed to understand the fundamentals of the concept and solve the exercise. Extra information should be left for the `after.md`, which is perfect for the content I've suggested removing.

Author: Moved to `after.md`.

> Kotlin has immutable (`val`) and mutable (`var`) variable. Immutable variable means that once it has it's value assigned - it can't be changed. Most of the time you will use this type of variable.

Reviewer: ```suggestion
Kotlin has immutable (`val`) and mutable (`var`) variables. The value of an immutable variable can't be changed after it's initial value is assigned. Most of the time you will use this type of variable.
```

```kotlin
val robotName = "HAL-9000"
userId = "T-1000"  // This will not compile
```

> Mutable variable means that it's value can be changed one or more times:

Reviewer: ```suggestion
A mutable variable's value can be changed one or more times:
```

```kotlin
var index = 12
print(index)  // 12

index = 100
print(index)  // 100
```

> BTW probably you have already noticed that semicolons in Kotlin are not used (they are optional but not needed except few special cases we will explore later in other exercises).

Reviewer: Maybe make this a bit more concise. Something like:

```suggestion
Semicolons in Kotlin are optional, except for a few special cases that will be covered later.
```

## Functions

> Functions in Kotlin are defined with keyword **`fun`** and are [first-class citizens][wiki-fcc] (not related to OOP). It means that you can declare (so-called `top-level functions`) them right in files (e.g. in Java you can define methods only in classes, not in files):

Reviewer: ```suggestion
Functions in Kotlin are defined with the `fun` keyword  and are _first-class citizens_ (not related to OOP). It means that you can declare (so-called `top-level functions`) them right in files (e.g. in Java you can define methods only in classes, not in files):
```

I've suggested a couple of changes:

- Removing the link to the wikipedia page. As per [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd): links should be used sparingly, if at all.
- Keywords (and code) are enclosed in backticks. See the [style guide](https://github.com/exercism/v3/blob/master/docs/maintainers/style-guide.md#code)

```kotlin
// This is content of the Hello.kt file

fun hello() {}
```

> Functions can receive arguments. Each argument have a name and a type. Probably you remember that compiler is smart enough to understand type of variable In many cases, but you always need to write types for function arguments by yourself as you are declaring certain contract. Functions can have zero or more arguments:

Reviewer: Maybe shorten this a bit to:

```suggestion
Functions can receive arguments. Each argument has a name and a type. Unlike variables, the type of arguments can't be inferred. Functions can have zero or more arguments:
```

```kotlin
fun hello() {}

fun hello(name: String) {}

fun hello(name: String, age: Int) {}
```

Kotlin functions might or might not return a value:

```kotlin
fun min(a: Int, b: Int): Int

fun countBonuses(user: User): Bonuses

fun run() {}
```

> As you can see, to return value from function you need to use keyword **`return`**:

Reviewer: ```suggestion
To return a value from a function, the `return` keyword is used:
```
